### PR TITLE
[WIP] Initial project structure

### LINKED ISSUE #1: Setup skeleton code with SWE best practices
- [x] #10
- [x] #5
- [x] #6
- [x] #7
- [x] #8
- [x] #9

### LINKED ISSUE #2: Setup components (backend, hardware, quantization, core)

### LINKED ISSUE #5: Unit tests in `pytest`

### LINKED ISSUE #6: Setup `pre-commit` (black, isort, pycln)
## Tools to include
- [x] black - code formatter
- [ ] mypy - static type checker
- [x] pycln - remove unused imports
- [x] isort - sort imports

## Skeleton to start from
```yaml
# See https://pre-commit.com for more information
# See https://pre-commit.com/hooks.html for more hooks
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.0.1
    hooks:
      - id: check-ast
      - id: check-builtin-literals
      - id: check-case-conflict
      - id: check-docstring-first
      - id: check-merge-conflict
      - id: check-json
      - id: check-toml
      - id: check-yaml
      - id: end-of-file-fixer
      - id: mixed-line-ending
      - id: trailing-whitespace
      - id: check-vcs-permalinks
      - id: check-shebang-scripts-are-executable
  - repo: https://github.com/pre-commit/pygrep-hooks
    rev: v1.9.0
    hooks:
      - id: python-check-mock-methods
      - id: python-no-log-warn
      - id: python-use-type-annotations
  - repo: https://github.com/hadialqattan/pycln
    rev: v1.2.5
    hooks:
      - id: pycln
        args: [--all]
  - repo: https://github.com/psf/black
    rev: 22.3.0
    hooks:
      - id: black
        additional_dependencies: ['click==8.0.4']
  - repo: https://github.com/PyCQA/isort
    rev: 5.10.1
    hooks:
      - id: isort
        args: ["--profile", "black"]
```

### LINKED ISSUE #7: Setup `ISSUE_TEMPLATE`
- [x] Feature request
- [x] Bug request

### LINKED ISSUE #8: Setup CI
- [x] ci.yml
- [ ] publish.yml
- [x] docs.yml
- [x] lint.yml

### ci.yml

```yaml
name: Run tests

on:
  push:
    branches:
      - main
      - develop
      - dev/*
    paths-ignore:
      - 'docs/**'
      - '**.md'
      - README.md
  pull_request:
    types:
      - opened
      - reopened
    branches:
      - main
      - develop
      - dev/*
    paths-ignore:
      - 'docs/**'
      - '**.md'
      - README.md

jobs:
  tests:
    runs-on: ubuntu-latest
    steps:
      ...
      - name: Checkout repository
        uses: https://github.com/actions/checkout/tree/v2v2
      - name: Run tests
        run: |
          echo "Running tests..."
          python -m pytest tests
```

### publish.yml
```yaml
name: publish_package_to_pypi

on:
  push:
    branches: ['main']

jobs:
  publish:
    runs-on: ubuntu-latest
    steps:
      ...
      - name: Publish Package
        run: |
          echo "Building package..."
          python -m poetry build
          echo "Publish package..."
          python -m poetry publish --username=${{ secrets.PYPIUSERNAME }} --password=${{ secrets.PYPIPASSWORD }}
```

### LINKED ISSUE #9: Setup `docs` in GitHub Pages
- [ ] Home: a simple introduction to the library and how to start to use it and installing it
- [ ] Tutorials: a set of tutorials to utilize the library
- [ ] Contribution: instructions on how to contribute

### LINKED ISSUE #10: Project Structure
The current suggested project structure

```
├── .github/
│   ├── workflows/
│   │   ├── ci.yml
│   │   ├── lint.yml
│   │   ├── docs.yml
│   │   └── publish.yml
│   └── ISSUE_TEMPLATE
├── approx/
│   └── core/
│       ├── backends
│       ├── hardware
│       └── quants
├── docs/
│   ├── docs.md
│   └── CONTRUBITING.md
├── examples/
│   ├── notebooks
│   └── scripts
├── tests
├── .gitignore
├── .pre-commit-config.yaml
├── LICENSE
├── README.md
└── pyproject.toml
```

### PR DESCRIPTION
**Pull request by bushshrub**
Closes: #1
Closes: #2
Closes: #5 
Closes: #6 
Closes: #7
Closes: #8
Closes: #9
Closes: #10

TODO:

- [x] #5 
- [x] #6 
- [x] #7 
- [x] #8 
- [x] #9 
- [x] #10

cc @sudomaze

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+site_name: approx
```
Comment (bushshrub) [inline mkdocs.yml:1]: Do we think mkdocs is the way to go? AFAIK, it's far easier to use as compared to sphinx.

File: approx/core/api.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    pass
+
+
+def auto_set_backend() -> None:
+    """
+    Automatically selects an appropriate backend to utilize.
+    Returns:
+        None
+    """
+    auto_select_backend()
```
Comment (bushshrub) [inline approx/core/api.py]: `auto_select_backend()` returns a BackendEngine but it is not being used?

File: approx/core/backends/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import torch
+
+from approx.core.backends.common import BackendEngine
+
+
+class PyTorchBackend(BackendEngine):
+    pass
```
Comment (bushshrub) [inline approx/core/backends/pytorch.py]: Should be implemented

File: mkdocs.yml (hunk 1 above)
Comment (ma7dev) [reply mkdocs.yml:1]: Yeah, let's use `mkdocs`. I have had my bad memories dealing with `sphinx` 😢

Comment (bushshrub) [reply mkdocs.yml:1]: Aight

**Review by ma7dev — APPROVED** [review]

**Pull request merged**